File: VB/CustomItemExtension/CustomItems/WebPageItem/WebPageItemMetadata.vb
Imports DevExpress.DashboardCommon
Imports System
Imports System.ComponentModel

Namespace DevExpresss.DashboardWin.CustomItemExtension
	<DisplayName("Web Page"), CustomItemDescription("Create a Web Page dashboard item and insert it into the dashboard." & vbLf & vbLf & "This item displays the web content. You can specify the URI pattern to create the page address from a data column at run-time."), CustomItemImage("WebPageItem.svg")>
	Public Class WebPageItemMetadata
		Inherits CustomItemMetadata

		<DisplayName("Attribute"), EmptyDataItemPlaceholder("Attribute")>
		Public Property Attribute() As Dimension
			Get
				Return GetPropertyValue(Of Dimension)()
			End Get
			Set(ByVal value As Dimension)
				SetPropertyValue(value)
			End Set
		End Property
		<UrlCustom>
		Public Property URI() As String
			Get
				Return GetCustomPropertyValue(Of String)()
			End Get
			Set(ByVal value As String)
				SetCustomPropertyValue(value)
			End Set
		End Property

		Public Class UrlCustomAttribute
			Inherits Attribute
			Implements ICustomPropertyValueConverterAttribute(Of String), ICustomPropertyHistoryAttribute(Of String)

			Public Function ConvertToString(ByVal value As String) As String Implements ICustomPropertyValueConverterAttribute(Of String).ConvertToString
				Return value
			End Function
			Public Function ConvertFromString(ByVal strValue As String) As String Implements ICustomPropertyValueConverterAttribute(Of String).ConvertFromString
				Return strValue
			End Function
			Public Function GetHistoryMessage(ByVal newValue As String, ByVal oldValue As String, ByVal dashboardItemName As String) As String Implements ICustomPropertyHistoryAttribute(Of String).GetHistoryMessage
				Return "URI for the " & dashboardItemName & " changed"
			End Function
		End Class
	End Class
End Namespace

File: CS/CustomItemExtension/CustomItems/WebPageItem/WebPageItemMetadata.cs
using DevExpress.DashboardCommon;
using System;
using System.ComponentModel;

namespace DevExpress.DashboardWin.CustomItemExtension
{
    [DisplayName("Web Page"),
    CustomItemDescription("Create a Web Page dashboard item and insert it into the dashboard.\n\nThis item displays the web content. You can specify the URI pattern to create the page address from a data column at run-time."),
    CustomItemImage("DevExpress.DashboardWin.CustomItemExtension.Images.WebPageItem.svg")]
    public class WebPageItemMetadata : CustomItemMetadata {
        [
        DisplayName("Attribute"),
        EmptyDataItemPlaceholder("Attribute"),
        ]
        public Dimension Attribute {
            get { return GetPropertyValue<Dimension>(); }
            set { SetPropertyValue(value); }
        }
        [UrlCustom]
        public string URI {
            get { return GetCustomPropertyValue<string>(); }
            set { SetCustomPropertyValue(value); }
        }

        public class UrlCustomAttribute : Attribute, ICustomPropertyValueConverterAttribute<string>, ICustomPropertyHistoryAttribute<string> {

            public string ConvertToString(string value) {
                return value;
            }
            public string ConvertFromString(string strValue) {
                return strValue;
            }
            public string GetHistoryMessage(string newValue, string oldValue, string dashboardItemName) {
                return "URI for the " + dashboardItemName + " changed";
            }
        }
    }
}
